### ISSUE TITLE
The registry already contains an instance for tl_page::id() when logging in via frontend

### ISSUE THREAD

**Issue by sleitz** [contao/core-bundle#995]
I encounter an exception when trying to log in via any frontend member in a Contao 4.4.2 installation  with several different member groups and members:

    RuntimeException:
    The registry already contains an instance for tl_page::id(95)

    at vendor/contao/core-bundle/src/Resources/contao/library/Contao/Model/Registry.php:182
    at Contao\Model\Registry->register(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Model.php:142)
    at Contao\Model->__construct(object(Result))
     (vendor/contao/core-bundle/src/Resources/contao/models/PageModel.php:761)
    at Contao\PageModel::findFirstActiveByMemberGroups(array(5))
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleLogin.php:100)
    at Contao\ModuleLogin->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:69)
    at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:483)
    at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:186)
    at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:227)
    at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:67)
    at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:422)
    at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:280)
    at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:177)
    at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:49)
    at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:306)
    at Contao\FrontendIndex->renderPage(object(Collection))
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:78)
    at Contao\FrontendIndex->run()
     (vendor/contao/core-bundle/src/Controller/FrontendController.php:42)
    at Contao\CoreBundle\Controller\FrontendController->indexAction()
    at call_user_func_array(array(object(FrontendController), 'indexAction'), array())
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:153)
    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:171)
    at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/app_dev.php:62)

The only meaningful online reference I found was this: contao/core#8443, but the circumstances are different. I tried to exclude some possible interference, mainly from extensions. But the stack trace also clearly shows no other bundles were involved. So this went nowhere.

What kind of information shall I provide in order for someone of you to understand this?

I can log in without failure in the online demo. But I have no idea where the differences are.

**Comment by sleitz** [comment]
By the way: This does not happen when using the frontend preview and emulating a member. It only happens in the real login process.

**Comment by leofeyer** [comment]
Does the issue occur in the online demo?

**Comment by sleitz** [comment]
Unfortunately not. But I do not know which specific circumstances of my instance could provoke it, so it might be reproducible given the right parameters.

**Comment by sleitz** [comment]
I invested some time now and found the following, based on the stack trace:

 * The member which tries to log in is member of one and only one member group
 * The member group is configured to redirect to a special page upon login of one of its members
 * The logic works as expected, the `ModuleLogin` takes the right turns in the if/else tree and asks the `PageModel` for the page to redirect to (in order to generate a URI from it)
 * The page model also works as expected, the database query is set up and the result count is exactly zero. In this specific case, page id 95 is returned as a complete database result
 * The result is passed to the page model constructor statically. the model itself has no constructor of its own, fall back to the generic `Model::__constructor()`
 * The construtor does nothing much, it follows some joined fields logic and loads the model registry
 * Saving into the model registry fails with the original exception message

I debugged the registry at this point. It contains the following entries:

 * `tl_page`: 25 items (which is not the complete tree, but very far away from zero), among them also id#95 (of course)
 * `tl_theme`: 1 item
 * `tl_module`: 14 items
 * etc.

I am no core developer, so I don't know about the whys and whens of the registry. I have no idea where the 25 page models already within the registry at that point originate from. But they are there.

I stumbled across one method that might help: If the `__construct` method would call `isRegistered()` before calling `register()`, this exception might not occur. Is that a viable solution? Or is it necessary to find out why the models gets registered beforehand?

**Comment by LarsLuz** [comment]
Hello!
I am from the same company as @sleitz . We are still experiencing this issue. Was there any progress that was not yet mentioned here? Is further data required to confirm / reproduce this problem?

Thank you in advance!

**Comment by leofeyer** [comment]
@contao/developers /cc

**Comment by Toflar** [comment]
> Or is it necessary to find out why the models gets registered beforehand?

Yes, the error message is actually correct. The case that you want to register a model to the registry that is already registered should never happen because you should fetch it from there.
As far as I can tell, `ModuleLogin` loads your page model ID `95` here: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/modules/ModuleLogin.php#L85. What you need to find out is who loaded `95` into the registry before and then **why** the `getRelated()` does not return the one that was already registered but tries to register it again instead.

**Comment by sleitz** [comment]
This is actually reproducible. Steps:

- Log into the online demo with k.jones
- Edit the member group "Violin Students"
- Check "Redirect on login"
- Select a page from the wizard, e.g. "Text elements", any will to
- Open a new private browser tab
- Log in to the online demo frontend with d.evans
- Experience the error message

**Comment by aschempp** [comment]
I can confirm the issue. The problem lies in `PageModel::findFirstActiveByMemberGroups`, because it uses it's own SQL query and not the `findBy` method.

**Comment by leofeyer** [comment]
We could fix it as follows:

```php

$objRegistry = \Model\Registry::getInstance();

/** @var PageModel|Model $objPage */
if ($objPage = $objRegistry->fetch('tl_page', $objResult->id))
{
	return $objPage;
}

return new static($objResult);
```

However, I wonder if this is correct? Why is there no `isRegistered()` check [here](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Model.php#L142)?

**Comment by Toflar** [comment]
> However, I wonder if this is correct? Why is there no isRegistered() check here?

I've already explained this. It's not correct to add that check there because it would make it even worse. This exception is thrown for a reason, we have to fix the source of the problem, not trying to make the exception disappear.
If you would add an `isRegistered()` check, your `findFirstActiveByMemberGroups()` would not throw an exception but instead return **a new model for the same page id** which will cause errors because suddenly there are 2 model instances for the same object so when a developer uses `->save()` on an instance, they will override each other (which exactly what the registry should prevent :)).

**Comment by leofeyer** [comment]
You are right, I remember now. 😄

**Comment by leofeyer** [comment]
Fixed in 504eb68dbda25d19096e3294a3ba2be291af0ea5 then.
